draft-ietf-quic-multipath.md

```diff
@@ -711,6 +711,26 @@ delay will most of the time relate to the path with the shortest latency.
 To collect ACK delays on all the paths, hosts should rely on time stamps
 as described in {{QUIC-Timestamp}}.
 
+If one chooses not to use time-stamps but wants to get reasonable estimation of
+RTTs on multiple paths with one packet number space, the following practices
+can be used:
+
+* The receiver side counts the number of ACK-eliciting packets received for each paths,
+  and keeps a per-path ACK timer. An ACK from that path is triggered when the
+  number of ACK-eliciting packets received on that path surpasses the path's
+  ACK-eliciting threshold or the path's ACK-timer expires.
+
+* Each path records the largest packet received on that path and the receive
+  time of that packet. When an ACK is sent on that path, the ACK Delay field is
+  re-calculated using that path's largest packet receive time.
+
+* The sender also keeps track of a list of sent packets for each path that are acknowledged by
+  ACKs received from the same path. A path's RTT sample is generated on receving ACK
+  that meets the following two conditions: (1) For that path, the largest
+  same-path acknowledged packet number is updated. (2) One of newly same-path
+  acknowledged packets is ACK-eliciting.
+
```

Comment:
I am trying to understand your algorithm. I guess it sums up as:

* For each path, the sender maintains knowledge of the largest PN acknowledged on path.
* The sender also keeps track of a list of sent packets for each path that are not yet acknowledged by
  ACKs received from the same path.
  
When a same path ACK is received:

* process the acknowledged ranges until finding a packet that:
      1) Was sent on the path
      2) Has a packet number larger than the largest PN acknowledged on path
* if several packets match the criteria above, retain only the one with the largest PN
* the remaining list may be empty, in which case do nothing
* if one packet matches the criteria above:
      1) Compute the RTT sample as the difference between the ACK received time and the packet sent time.
      2) Update RTT average, etc.
      3) Set the largest PN acknowledged on path to the PN that was just processed
      3) Remove all packets with PN lower or equal than this largest acknowledged PN from the list of sent packets for each path that are not yet acknowledged by ACKs received from the same path

The downside of that method is that the RTT sample may include some or all of the ACK Delay of the ACK frame. The possible rule is:

* if the largest acknowledged number in the ACK frame matches the largest acknowledged number on path, remove the ACK Delay from the RTT sample.
* if not, accept that the measurement may be imprecise.

Do I understand correctly?

Reply:
Of course, we need to also specify the receiver behavior. Specifying that behavior may help deal with the "ACK Delay" issue. I assume that this why you write:

* The receiver counts the number of ACK-eliciting packets received for each paths,
  and keeps a per-path ACK timer. An ACK from that path is triggered when the
  number of ACK-eliciting packets received on that path surpasses the path's
  ACK-eliciting threshold or the path's ACK-timer expires.

I am not too sure about that, because the net effect of that rule is to get at least one new ACK **for each path** per ACK Timer. If the goal is to provide accurate measurement of RTT, we don't need that many "on path ACK". One per RTT would be enough. Maybe 2 per RTT to be safe. Or the max value of "ACK Timer" and "Path RTT / 2". Reset that timer for each path after sending an on path RTT. Then, trigger an on path ACK as soon as (1) the timer has expired and (2) a packet has been received. (Ack eliciting or not does not matter.)

That way, the sender can assume that for the on-path packet, the ACK delay is set to zero.

Reply:
Correction. My concerned about ACK Delay is misplaced. We are speaking about ACK_(MP), so the largest packet acknowledged is always a packet sent on that same path. See the modification that I propose on PR #129, a simple guidance for always sending at least some ACK on the same path. If we accept that guidance, then the text becomes simpler -- no need to specify a receiver side strategy. So the entire addition becomes:


* For each path, the sender maintains knowledge of the largest PN acknowledged on path.

* The sender also keeps track of a list of sent packets for each path that are not yet acknowledged by ACKs received from the same path.

* When a same path ACK is received, the sender checks whether the largest packet acknowledged is larger than the largest PN acknowledged on path. If it is, the sender will:

    - Compute a new path RTT sample using the ACK arrival time, the ACK delay indication, and the time at which the packet with this largest PN was sent.
    - Set the largest PN acknowledged on path to the PN that was just processed
    - Remove all packets with PN lower or equal than this largest acknowledged PN from the list of packets sent on this path that are not yet acknowledged by ACKs received from the same path.

Reply:
Of course, there is the issue of packets sent with zero-length CID. In that case, the simple algorithm above does not work very well. The ACK_(MP) will contain ack ranges received on many paths. Instead of using the largest number in the ACK_(MP) frame, the sender needs to filter the ACK ranges and retrieve the largest number acknowledged for the path.